test/vmtests/vmtests/utils/local_client.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Mostly just a wrapper around the subprocess library but can
+# live log stdout and stderr, while also collecting them as
+# strings.
+
+import logging
+import subprocess
+import time
+from io import StringIO
+from pathlib import Path
+from threading import Thread
+from typing import IO, Any, Dict, List, Optional, Union
+
+
+# The result of a subprocess execution.
+class LocalExecutableResult:
+    def __init__(
+        self,
+        stdout: str,
+        stderr: str,
+        exit_code: int,
+        cmd: Union[str, List[str]],
+        elapsed: float,
+        is_timeout: bool,
+    ) -> None:
+        self.stdout = stdout
+        self.stderr = stderr
+        self.exit_code = exit_code
+        self.cmd = cmd
+        self.elapsed = elapsed
+        self.is_timeout = is_timeout
+
+    def check_exit_code(self) -> None:
+        if self.is_timeout:
+            raise Exception("Process timed out")
+
+        elif self.exit_code != 0:
+            raise Exception(f"Process failed with exit code: {self.exit_code}")
+
+
+# Handles reading a pipe (stdout or stderr).
+# The contents are both collected as a string and logged.
+class _PipeReader:
+    def __init__(self, pipe: IO[str], log_level: int, log_name: str) -> None:
+        self._pipe = pipe
+        self._log_level = log_level
+        self._log_name = log_name
+        self._output: Optional[str] = None
+
+        self._thread: Thread = Thread(target=self._read_thread)
+        self._thread.start()
+
+    def close(self) -> None:
+        self._thread.join()
+
+    def __enter__(self) -> "_PipeReader":
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        self.close()
+
+    def wait_for_output(self) -> str:
+        self._thread.join()
+
+        assert self._output is not None
+        return self._output
+
+    def _read_thread(self) -> None:
+        log_enabled = logging.getLogger().isEnabledFor(self._log_level)
+
+        with StringIO() as output:
+            while True:
+                # Read output one line at a time.
+                line = self._pipe.readline()
+                if not line:
+                    break
+
+                # Store the line.
+                output.write(line)
+
+                # Log the line.
+                if log_enabled:
+                    line_strip_newline = line[:-1] if line.endswith("\n") else line
+                    logging.log(self._log_level, "%s: %s", self._log_name, line_strip_newline)
+
+            self._pipe.close()
+            self._output = output.getvalue()
+
+
+class LocalProcess:
```

Comment:
Are we writing a new framework for running processes or is it being ported? Is this related to LISA in anyway? (I don't have context - but probably we have similar code somewhere)

Reply:
This is new code. But the API was inspired by the [LISA node](https://github.com/microsoft/lisa/blob/main/lisa/node.py#L244) type. Though, I did simplify some of the params. 

Having stderr and stdout be automatically logged while also collecting them as strings is incredibly useful. It makes debugging test failures a lot easier. Hence, I copied the idea.

In a future change, I will add a very similar API for SSH.